### PR TITLE
Script to install dependencies (except Java)

### PR DESCRIPTION
**Pull request by mbudiu-vmw**
I also changed Spec.hs to preserve the PATH, otherwise it won't find javac if it is installed in a non-system directory.

### CODE REVIEW

File: install-dependencies.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+echo "Installing dependencies for DDlog"
+echo "This script should be invoked with '. ./install-dependencies.sh' to set up the environment properly"
+
+case "$OSTYPE" in
+    linux*) ;;
+    *) echo "Unhandled operating system $OSTYPE"; exit 1;;
+esac
+
+
+echo "Installing Haskell"
```
Comment (ryzhyk) [inline install-dependencies.sh]: There is a script to install Haskell in `.travis/install-stack.sh`, and it works on both Linux and MacOS.  Shall we call that instead?

File: install-dependencies.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+curl -sSL https://get.haskellstack.org/ | sh
+
+echo "Installing Rust"
+curl https://sh.rustup.rs -sSf | sh -s - -y
+export PATH=$HOME/.cargo/bin:$PATH
+
+echo "Installing Flatbuf"
```
Comment (ryzhyk) [inline install-dependencies.sh]: There is a script to install FlatBuffers in `.travis/install-flatbuf.sh`.  It works for both MacOS and Linux. Only problem is, it installs in the home directory (which makes sense on Travis, but not necessarily on the developer's machine). Your script, on the other hand, installs in the `differential-datalog` directory. I can probably change the Travis script to do the same, so we only have one script to install Travis.

File: test/Spec.hs [hunk 3]
```diff
@@ -192,9 +192,11 @@ compilerTest progress java fname cli_args crate_types = do
         compileFlatBufferBindings prog specname (dir </> rustProjectDir specname)
     -- compile generated Java code
     classpath <- (maybe "" (":" ++ )) <$> lookupEnv "CLASSPATH"
+    path <- (maybe "" (":" ++ )) <$> lookupEnv "PATH"
```
Comment (ryzhyk) [inline test/Spec.hs]: Can be just `path <- (maybe "" id) <$> lookupEnv "PATH"`, as we are not appending anything to it.

File: install-dependencies.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+# Script that installs the dependencies needed to build and run Differential Dataflow
+
+echo "Installing dependencies for DDlog"
+echo "This script should be invoked with '. ./install-dependencies.sh' to set up the environment properly"
+
+case "$OSTYPE" in
```
Comment (ryzhyk) [inline install-dependencies.sh]: We can totally auto-install dependencies on MacOS (see below)

File: README.md [hunk 5]
```diff
@@ -67,7 +67,11 @@ Datalog 2.0, Philadelphia, PA, June 4-5, 2019.
 
 ## Installing from sources
 
+To install the required dependencies for buidling DDlog run `. ./install-dependencies.sh`
```
Comment (ryzhyk) [inline README.md]: The comment should mention that manual installation is required if the user has Haskell or Rust or FlatBuffers already installed and want to keep their existing versions. Alternatively we need more complex logic to handle existing compilers.

File: install-dependencies.sh (hunk 1 above)
Comment (mbudiu-vmw) [reply install-dependencies.sh]: The travis script depends on a lot of Travis-related stuff. 
This command-line is also very short, much shorter than your script.
I should investigate how to do it for the Mac, but I can't really test.

File: install-dependencies.sh (hunk 2 above)
Comment (mbudiu-vmw) [reply install-dependencies.sh]: There you are worrying about travis retrying too.
It's short enough that some duplication should be OK.

Comment (mbudiu-vmw) [reply install-dependencies.sh]: Your scripts for Travis are very different from the manual installation instructions you are giving in the README.

File: install-dependencies.sh (hunk 1 above)
Comment (ryzhyk) [reply install-dependencies.sh]: There's nothing travis-related in the script it literally just checks OS type and runs `curl`. Despite its name, `travis_retry` is declared in the same file and simply retries the command several times.

File: tools/install-flatbuf.sh [hunk 6]
```diff
@@ -2,25 +2,26 @@
...
+    echo "Installing Flatbuf"
+    retry fetch_flatbuf
     cd flatbuffers
     cmake -G "Unix Makefiles"
     make
+    cd ..
 fi
 
-ln -f -s $HOME/flatbuffers/flatc ~/.local/bin/flatc
+ln -f -s ./flatbuffers/flatc ~/.local/bin/flatc
```
Comment (ryzhyk) [inline tools/install-flatbuf.sh]: You don't want to use relative paths in symlinks; should probably be:
```
ln -f -s `pwd`/flatbuffers/flatc ~/.local/bin/flatc
```

File: .travis.yml [hunk 7]
```diff
@@ -61,7 +61,7 @@ before_install:
 - export PATH=$HOME/.local/bin:$PATH
 - ./tools/install-stack.sh
 - ./tools/install-flatbuf.sh
-- export CLASSPATH=$CLASSPATH:$HOME/flatbuffers/java
+- export CLASSPATH=$CLASSPATH:./flatbuffers/java
```
Comment (ryzhyk) [inline .travis.yml]: Likewise, we should not use relative `$CLASSPATH`:
```
export CLASSPATH=$CLASSPATH:`pwd`/flatbuffers/java
```

**Comment by ryzhyk** [comment]
Also there's a build failure:
```
/Users/travis/build/vmware/differential-datalog/test/Spec.hs:196:5: error: [-Wname-shadowing, -Werror=name-shadowing]

    This binding for ‘path’ shadows the existing binding
      imported from ‘Test.HUnit’ at test/Spec.hs:34:1-17
      (and originally defined in ‘Test.HUnit.Base’)
    |
196 |     path <- (maybe "" id) <$> lookupEnv "PATH"
    |     ^^^^
```

File: .travis.yml (hunk 7 above)
Comment (ryzhyk) [reply .travis.yml]: Do you want to fix this one too (otherwise, I can do it; just want to be sure you're ok with the suggestion)?

Comment (mbudiu-vmw) [reply .travis.yml]: I actually fixed this one too, but I mistyped the file name when I committed...
Sorry about that. Pushed it now.

**Pull request merged**